sklearn/naive_bayes.py

```diff
@@ -958,3 +998,256 @@ def _joint_log_likelihood(self, X):
+            assumed to be from a different categorical distribution.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        classes : array-like, shape = [n_classes] (default=None)
+            List of all the classes that can possibly appear in the y vector.
+
+            Must be provided at the first call to partial_fit, can be omitted
+            in subsequent calls.
+
+        Returns
+        -------
+        self : object
+        """
+        return super(CategoricalNB, self).partial_fit(X, y, classes,
+                                                      sample_weight=None)
+
+    def _check_input(self, X, Y=None):
+        if Y is None:
+            if _old_numpy:
+                X = check_array(X, accept_sparse=False)
+                if not np.isfinite(X).all():
+                    raise ValueError("Input contains NaN or infinity.")
+                return np.asarray(X, dtype=np.int64)
+            else:
+                return check_array(X, accept_sparse=False)
+        else:
+            if _old_numpy:
+                X, Y = check_X_y(X, Y, accept_sparse=False)
+                if not np.isfinite(X).all():
+                    raise ValueError("Input contains NaN or infinity.")
+                X = np.asarray(X, dtype=np.int64)
+                return X, Y
+            else:
+                return check_X_y(X, Y, accept_sparse=False)
+
+    def _init_counters(self, n_effective_classes, n_features):
+        self.class_count_ = np.zeros(n_effective_classes, dtype=np.float64)
+        self.category_count_ = [np.zeros((self.class_count_.shape[0], 0))
+                                for _ in range(n_features)]
+        self.feature_cat_index_mapping_ = [{} for _ in range(n_features)]
+
+    def _count(self, X, Y):
+        if np.any((X.data if issparse(X) else X) < 0):
+            raise ValueError("Input X must be non-negative")
+        self.class_count_ += Y.sum(axis=0)
+        for i in range(self.n_features_):
+            X_feature = X[:, i]
+            categories = np.unique(X_feature)
+            self._update_mapping(self.feature_cat_index_mapping_[i],
+                                 categories)
+            # update category_count_dimensions in case partial_fit is used
+            self.category_count_[i] = self._update_category_count_dimensions(
+                self.category_count_[i], self.feature_cat_index_mapping_[i])
+            self._update_category_count(X_feature, Y,
+                                        self.feature_cat_index_mapping_[i],
+                                        self.category_count_[i])
+
+    def _count_categories(self, X_feature):
+        if _old_numpy:
+            bins = np.bincount(X_feature)
+            non_zero_bins = np.nonzero(bins)[0]
+            return non_zero_bins, bins[non_zero_bins]
+        else:
+            return np.unique(X_feature, return_counts=True)
+
+    def _update_mapping(self, cat_mapping, categories):
```

Comment:
Nice, really cool trick!
I just checked the code. Unfortunately, in my opinion, to really benefit from the increased ease of use, some non-trivial code changes would be necessary.
As of now, `category_count` for a respective feature is a numpy array with shape (num_classes, num_cats). When `partial_fit` is used, the (shape of the) numpy array needs to be updated, in order to account for unseen categories. For this, we have to get all unique categories for this `partial_fit` and combine them with the old categories, to see what has changed.  So, before we can benefit from the new dictionary and add/access the new values "on the fly" we have to use the values already, to update the data structure.
So, all in all,  we would save one line of code but would risk making the code a little harder to understand in case the reader is not familiar with `defaultdict.default_factory` and we have to be more careful with potential KeyErrors, which makes the code a little harder to maintain.
The alternative would be, to choose a new data structure for `category_count`. E.g. something like choosing a list of Counters. However, this could slow down the calculations.
So, even though I really like the idea, I think the disadvantages outweigh the benefits.

Reply:
@timbicker, I don't exactly understand why you would need to change the data structure of `category_count` for that. Looking at the code I think you could just use @jnothman's trick and get rid of `_update_mapping`. Neither in `_update_category_count_dimensions` nor `_update_category_count` you seem to modify `self.feature_cat_index_mapping_[i]`.

Reply:
@FlorianWilhelm Thanks for your reply.
Yes, you are right that neither `_update_category_count_dimensions` nor `_update_category_count` modify `self.feature_cat_index_mapping_[i]`. However, they need the length of the already updated mapping, to update the count dimensions. This means I have to do some sort of mapping update (or at least extract all new categories from `X`) before we can call these functions.
The ideal use case would be, imo, to update the mapping automatically inside`_update_category_count` via `indices = [cat_mapping[category] for category in class_cats]` without having to change anything. But, we cannot get totally rid of some sort of mapping update or category extraction. So why not do a manual update of dict, when we have to get the new categories anyways and make the update explicit to the user/developer?

Since, in my opinion, the total benefit of `default_dict` is, in this case, rather small. I, personally, would leave it like that. But maybe I just see things to problematic and I should just change it. Let me know :)